spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -814,8 +819,17 @@ struct ProductionVisitor
             pb->saveParsePosition();
         }
 
-        if ( ncur )
+        if ( ncur ) {
+            // For `&max-size` we use `ncur` to store a view limited and slightly larger than the _permissible_ range,
+            // not the _consumed_ range. We need to manually compute the actually consumed range from where we ended
+            // parsing in the limited view. We update the value stored in `ncur` so it can be updated below.
+            if ( AttributeSet::find(field->attributes(), "&max-size") )
+                ncur = builder::memberCall(state().cur, "advance",
+                                           {builder::difference(builder::memberCall(*ncur, "offset", {}),
+                                                                builder::memberCall(state().cur, "offset", {}))});
+
```

Comment:
Unfortunately this is needed. For `&max-size` (also for `&size`) `ncur` is into the the limited view while the `cur` here on the full input; the`popState` transitioning between them is https://github.com/zeek/spicy/blob/8044db245af27136838a7793b1ec5445c4fd8ddf/spicy/toolchain/src/compiler/codegen/parser-builder.cc#L813

By stashing away the limited view in `ncur` for `&max-size` we have access to it here and can referenced it in offset computations, but bemoving this block here would lead to us pointing into the limited view which is incorrect (previously we'd point after the slightly larger limited view, now we'd hypothetically point into the limited view so we could only consume the sentinel byte).

Reply:
Ok, thanks for digging in. Then how about at least not using `ncur` to stash away the earlier `cur`, but some new temporary variable that we then use here to now set `ncur`. That should make it a bit easier to understand this.

Reply:
I introduced a new variable to track this state without overloading `ncur` (suggestions on naming welcome).